homeassistant/components/scheduler/__init__.py

```diff
@@ -0,0 +1,506 @@
+        """Schedule the next instance of a schedule."""
+        now = datetime.now()
+
+        # Calculate the start and end datetimes for the next instance:
+        if not self._initial_run_complete:
+            self._initial_run_complete = True
+            start_datetime = self.start_datetime
+            end_datetime = self.end_datetime
+        elif self.rrule:
+            start_datetime = self.rrule.after(now, inc=True)
+            if self.end_datetime:
+                end_datetime = start_datetime + self._duration
+            else:
+                end_datetime = None
+
+        if not start_datetime:
+            _LOGGER.debug("No more instances of schedule: %s", self)
+            return
+
+        self.active_instance: ActiveScheduleInstance = ActiveScheduleInstance(
+            self._hass,
+            self.async_schedule_next_instance,
+            self.entity_id,
+            start_datetime,
+            end_datetime=end_datetime,
+        )
+        self.active_instance.async_init()
+
+        _LOGGER.info(
+            "Scheduled instance of schedule %s (start: %s / end: %s)",
+            self,
+            start_datetime,
+            end_datetime,
+        )
+
+
+@bind_hass
+class Scheduler:
+    """A class to manage scene schedules."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize."""
+        self._hass: HomeAssistant = hass
+        self._latest_instances: List[ActiveScheduleInstance] = []
+        self._store = Store(hass, STORAGE_VERSION, DOMAIN)
+        self.schedules: Dict[str, Schedule] = {}
+
+    async def async_clear_expired_schedules(self) -> None:
+        """Delete expired schedules."""
+        now = datetime.now()
+        self.schedules = {
+            schedule_id: schedule
+            for schedule_id, schedule in self.schedules.items()
+            if (not schedule.end_datetime and schedule.start_datetime < now)
+            or (schedule.end_datetime and schedule.end_datetime < now)
+        }
+
+        await self.async_save()
+
+    async def async_create(
+        self,
+        entity_id: str,
+        start_datetime: datetime,
+        *,
+        end_datetime: Optional[datetime] = None,
+        rrule_str: Optional[str] = None,
+        schedule_id: Optional[str] = None,
+    ) -> dict:
+        """Create a schedule.
+
+        The rrule_str parameter must be an RFC5545-compliant string.
+        """
+        if not schedule_id:
+            schedule_id = uuid4().hex
+
+        schedule = Schedule(
+            self._hass,
+            schedule_id,
+            entity_id,
+            start_datetime,
+            end_datetime=end_datetime,
+            rrule_str=rrule_str,
+        )
+        self.schedules[schedule_id] = schedule
+
+        await self.async_save()
+        _LOGGER.info("Created schedule: %s", schedule)
+        await schedule.async_schedule_next_instance()
+
+        return schedule.as_dict()
```

Comment:
This makes no sense as part of a create method. Make sure you don't conflate websocket command responsibilities and this class.